rtio: executor: implement RTIO_OP_AWAIT in executor

Emil-Juhl opened the pull request:
Allow the rtio executor to handle the RTIO_OP_AWAIT much like how it can take care of the DELAY_OP.

The RTIO_OP_AWAIT is similar to the DELAY_OP in the sense that it just makes the rtio execution pause until the RTIO_OP_AWAIT is completed. Instead of tying this completion to a fixed time delay, the RTIO_OP_AWAIT requires the user to signal completion using the corresponding rtio_sqe. After this, the rtio execution will continue.

A default rtio_executor implementation is useful for device drivers that use some kind of ready flag during transactions. For example in order to read data out of a sensor, the following flow may be required:
    1. Write some cmd like "read accelerometer data"
    2. Await data-ready GPIO rising edge
    3. Read the requested data payload from the sensor Using the RTIO_OP_AWAIT this can be elegantly put together in a single chained rtio operation.

Of course, any rtio_iodev drivers may provide an alternative implementation wherein additional reasonable operations are added during the waiting period. As an example, a driver could block all bus transactions during the waiting period, if that seems reasonable to do.

Emil-Juhl committed:
rtio: executor: add relaxed RTIO_OP_AWAIT variant

Allow the rtio executor to handle the RTIO_OP_AWAIT for cases where no
rtio_iodev needs to be blocked while waiting for the signal.

When the RTIO_OP_AWAIT was introduced, the highlighted usecase was
performing transactions within an exact time window. This can be
achieved by blocking the rtio_iodev, effectively locking the bus, while
awaiting.

However, a relaxed implementation which doesn't block any rtio_iodev is
really useful for device drivers that use some kind of ready-event
during transactions. For example in order to read configuration out of
a sensor, the following flow may be required:

    1. Write some cmd like "read accelerometer scale"
    2. Await data-ready GPIO rising edge (might take "a long time")
    3. Read the requested data payload from the sensor

Using a relaxed variant of the RTIO_OP_AWAIT this can be elegantly put
together in a single chained rtio sequence. And by not blocking the
rtio_iodev, other devices can be operated during the await period.
Of course, not blocking the rtio_iodev also means that the bus may be
busy when the RTIO_OP_AWAIT completes and thus, step 3 may be blocked
for a some time and the guarantee of an exact time window cannot be
achieved with the relaxed variant.

To make the API clearer, separate helpers are added for the strict
(blocking iodev) and relaxed (blocking only the sqe chain) variant
respectively.

Signed-off-by: Emil Dahl Juhl <emil@s16s.ai>

bjarki-andreasen reviewed (changes requested):
Breaks intent of the RTIO await operation

Emil-Juhl committed:
tests: rtio: rtio_api: add relaxed await sqe test

Expand the test case for RTIO_OP_AWAIT to also cover the relaxed variant
where just the SQE chain is paused until completion, and the rtio_iodev
can serve unrelated SQEs meanwhile.

Signed-off-by: Emil Dahl Juhl <emil@s16s.ai>

Emil-Juhl commented:
> ok, I did not see the path where the iodev is NULL,
>
Then I completely understand the confusion.
> it did seem strange that the addition in this PR passed the rtio test suite as I specifically added a case to ensure the iodev is blocked. Could you add docs to this API
> https://github.com/zephyrproject-rtos/zephyr/blob/101add04bbb56a1f1e36f7cec612463ea7a3e399/include/zephyr/rtio/rtio.h#L748-L758
> making it clear that the iodev can be NULL, and that whether it is passed or not controls if it blocks the iodev or the executor? 
>
I decided to try to make the API clear by adding a separate prep function for what I called "relaxed" await. Then through an ASSERT on the original prep function, and docs for both, I hope that the consequences of choosing one over the other is quite clear.

> and also extend the test suite to test blocking the executor?
>
Yeah, I also wrote another test function within the existing test case to test that the relaxed variant allows a separate rtio sqe to complete while the await is "in progress".

By all means let me know if you still have remarks for this approach

Emil-Juhl commented:
> As an aside, it would be great to see an await event type OP. In theory this is whats missing for things like programming the event -> action combos that many many hardware vendors provide for basically offloading everything but optionally catching a DMA completion in software.
> 

I could totally see this being useful, yes.

> One of the hassles with this tends to be how do you exclusively take the hardware, but since we are already dealing with this now (await with an iodev and transaction) its maybe possible to say (await event X with an iodev and transaction) to be the thing that a driver could look at, say ok cool we have _hardware_ for this and program it accordingly.
> 

I imagine it could prove difficult to map out sort-of cross-domain events together. The thing that RTIO_OP_AWAIT currently dealt with was locking a specific bus until the sqe got signalled. But mapping that to e.g. an interrupt on the gpio still functionally ends up as a "software" event. I could see that through some descriptive fields, the RTIO_OP_AWAIT could specify some event to await, like a specific gpio interrupt, but I am not quite sure how the respective rtio_iodev driver would configure the system for the purpose.
From what I've seen so far, basically the rtio_iodev will represent a bus, like SPI. But for this type of "await a HW event" wouldn't the rtio_iodev need to either:
1. Know about the HW platform on a deeper level to reconfigure and route things as necessary in order to offload as much as possible to the hardware
2. Be able to use some APIs in order to poke the other relevant drivers, e.g. a gpio driver, to provide a specific configuration.

To me, case 2 sounds like the "right way" to approach the problem. Both in terms of architecture but also general feasibility. Maybe the existing APIs are actually already more or less what would be needed, but I'm not sure if it will go all the way to the ideal maximum amount of offloading.

It's not really a thing that I would be looking a lot into myself, though it's an interesting problem. But since you brought it up, I at least wanted to give my two cents back. Though consider that I am not deeply familiar with all of this at all.

> The event in the current await op happens to be a _software_ event. But it'd be super neat to see a _hardware_ event here. @PetervdPerk-NXP for sure I believe would be interested in seeing this with the crossbar and edma. Perhaps @MaureenHelm would be interested in this for high performance ADCs on or off chip given the max32 and large selection of adc/dac parts Analog has.

Emil-Juhl commented on include/zephyr/rtio/rtio.h:
```diff
@@ -745,18 +745,48 @@ static inline void rtio_sqe_prep_transceive(struct rtio_sqe *sqe,
+ *
+ * See @ref rtio_sqe_prep_await_relaxed for a variant, where no rtio_iodev is blocked.
+ */
 static inline void rtio_sqe_prep_await(struct rtio_sqe *sqe,
```
Alright, I just initially figured that maybe having three options where one of them can be serve as either of the others would potentially cause confusion for users.

I have implemented the two helpers and kept the existing prep_await as-is now.

Emil-Juhl commented on include/zephyr/rtio/rtio.h:
```diff
@@ -745,18 +745,48 @@ static inline void rtio_sqe_prep_transceive(struct rtio_sqe *sqe,
+ * This relaxed variant can be useful when the await op serves as a logical piece of a sequence
+ * without requirements for a low-latency continuation of the sequence upon completion.
+ */
+static inline void rtio_sqe_prep_await_relaxed(struct rtio_sqe *sqe, int8_t prio, void *userdata)
```
Should be applied.

Emil-Juhl commented on include/zephyr/rtio/rtio.h:
```diff
@@ -580,7 +580,7 @@ struct rtio_iodev {
 #define RTIO_OP_I3C_CCC (RTIO_OP_I3C_CONFIGURE+1)
 
-/** An operation to suspend bus while awaiting signal */
+/** An operation to await a signal, while potentially suspending the iodev, before completing */
```
Yeah, that's a better phrasing.

bjarki-andreasen reviewed (dismissed):
Love it! really nice work

Emil-Juhl commented on tests/subsys/rtio/rtio_api/src/test_rtio_api.c:
```diff
@@ -949,9 +949,82 @@ void test_rtio_await_(struct rtio *rtio0, struct rtio *rtio1)
+
+	/* Submit the rtio0 sequence and make sure it reaches the AWAIT sqe */
+	TC_PRINT("Submitting await sqe from rtio0\n");
+	res = rtio_submit(rtio0, 0);
```
You're right that it does work. It's only the specific chain which is "blocked" during the await. I will update the test to cover this as well. And rephrase commit messages and comments to avoid talking about "blocking the executor".
For the intended use of the _executor variant of await, this is perfectly fine.

However, it does open another question with regards to the strict variant. I just tried to plonk the same test into the strict variant and what I get is basically:
```
1. NOP chained completes
2. NOP freestanding completes
3. AWAIT completes (when signaled, of course)
```
Now, I understand that when it comes to the rtio_iodev implementation, the test uses a different one than say the one in `i2c_rtio.c`. However, a lot of the code seems similar to me. At least in terms of the `submit` and `_next` functions.
So whether or not the physical bus is blocked during the await is of course a detail within specific drivers (like `nrfx_twim_rtio`) but what I am not so certain about is the potential inversion of order in the sqe queue?
Given the intention of the strict AWAIT op, I would expect the above test to result in:
```
1. NOP chained completes
2. AWAIT completes (when signaled)
3. NOP freestanding completes (only _after_ the chained AWAIT above)
```
Since if the freestanding NOP can make its way in-between parts of the chained operation, then the guarantee of a low-latency continuation after the AWAIT can't really be given, right? The NOP _could_ be a bus operation that will take some time to complete, or maybe even be delayed until after the AWAIT if the bus is truly locked in which case it would kind of land on the worst time.

I guess it's quite out of scope for this PR, but it's an interesting issue. Or maybe (hopefully) I am mistaken and it's not really a problem at all.

bjarki-andreasen replied:
ok, the executor itself is not blocked, only the one chain, that makes sense (I did not know we already supported this, its nice), are we aligned though that an iodev would be blocked by the await, as its an SQE which does not finish until signaled, so before the iodev informs the executor the SQE has been finalized, it can not continue?

Emil-Juhl replied:
> [...] are we aligned though that an iodev would be blocked by the await, as its an SQE which does not finish until signaled, so before the iodev informs the executor the SQE has been finalized, it can not continue regardless if whether its part of a chain or not?
>
I agree with this and I believe it matches my observation when I modified the test case.

The potential problem that I see is that there doesn't seem to be a guarantee that the next SQE to actually be worked on will be the next-in-chain from the AWAIT operation. At least, when I just tried in the test case (keep in mind the rtio_iodev implementation differs from e.g. the i2c_rtio.c) it seemed to be the case that the "freestanding" SQE would be executed just after the AWAIT completes, and thus _before_ the next-in-chain.

But maybe I am just misunderstanding the use a bit, and that to get this guarantee in practice you must use RTIO_SQE_TRANSACTION rather than just RTIO_SQE_CHAINED?
